source/FreeRTOS_DNS.c

```diff
@@ -557,7 +557,7 @@
              * is being used, so the buffer must be freed by the
              * task. */
             FreeRTOS_ReleaseUDPPayloadBuffer( xReceiveBuffer.pucPayloadBuffer );
-        } while( 0 );
```

Comment:
Either use a goto or multiple returns and add a deviation (potentially project-wide). This is just plain confusing.

Reply:
@paulbartell : a do-while-false-loop is used in **many situations** throughout the +TCP library, and I'd like to keep them.

> Either use a goto

The use of GOTO's would work like a red flag and will gather less approval. And beside that, I think it was MISRA or lint that doesn't like them.

> or multiple returns

MISRA "forbids" them :-(

I recommend to leave these loops in the source code for now.

We should avoid making big aesthetic changes to the code, because that makes it much harder to compare the source code **through time**.
Since the renaming of files, and since the splitting-up in smaller files, I already have a hard time making sure that code it OK in all respects.

One of the things that I try to do is find lost PR's: that is a change in the main branch that got lost while we used a temporary branch "IntergrationTestingX".

Another thing I do is testing the libraries "in the field".

Reply:
To summarize the relevant MISRA rules:

Rule 15.1: Advisory: The goto statement shall not be used.

Rule 15.4: Advisory: There should be no more than one break or goto statement used to terminate any iteration statement.

Rule 15.5: Advisory: A function should have a single point of exit at the end.

Advisory level rules just need some kind of justification... It's not a "ban".

At a high level, do-while-false, goto / label, and multiple returns all violate the underlying principle... jumping over code can lead to unwanted side effects.

Reply:
I must say that the while(0) break pattern is identical in substance if not form to the goto and one exit/function advisories.  The idea is to comply to the spirit of the MISRA rules and not simply the specifics.

That said, the MISRA rule against goto 15.1 is advisory as Paul points out and importantly, rule 15.2: states that any goto's must jump to a label in the same function.  Following the goto/cleanup exception C idiom is probably a better way in the long run.  BUT I agree that the while(0) break idiom is the one used in +TCP and consistency is important.